Let me explain our application and why we need this feature.

cpinto-cb opened the pull request:
We search data from network devices and the data from these devices is stored in elasticsearch. A search (filter) query can be sent to find information (terms/keywords), but elasticsearcg also has the ability to say “not these terms” in a query as well. As such, our UI would be be designed with a query filter input and a "NOT these terms" filter input. So if you will, two global filter inputs. One input for “find this stuff” and the other input for “but not this stuff” or exclude this stuff in the results.

I have made changes to the code so that this information comes through to the server data function like so…

DEVICEREPORTS.setElasticFnServerData = function (type, indice_pattern, deviceCustomString2) {

    var recordType = type,
        service = deviceCustomString2;

    this.setFnServerData(function (sSource, aoData, fnCallback, oSettings) {
        var body,
            builder,
            draw = oSettings.iDraw,
            length = oSettings._iDisplayLength,
            start = oSettings._iDisplayStart,
            globalSearch = oSettings.oAjaxData.search.value,
            globalExclude = oSettings.oAjaxData.exclude.value,
            columnSearches = aoData[1].value,

These changes add the functionality needed for a server data function to work.
I need to learn more about the code base to make this feature work client side too.

DataTables commented:
Thanks for the pull request - much appreciated!

There are quite a lot of white space changes so I wouldn't be able to take it as it is. That also makes the diff needlessly large and its hard to see where the actual changes are. My guess is that your editor is automatically changing tabs to spaces.

I'm also not clear why this needs to be in DataTables core and not added as a plug-in. The [feature plug-ins](https://datatables.net/manual/plug-ins/features) should provide everything needed for this. You can register the DOM element and a filter plug-in that would perform the exclude action.

Before reviewing the code any further I'd like to just seek some clarity on that point.

Thanks!

cpinto-cb commented:
Allan,

If there is a way to get data to the server function that would be great.
I looked at one of the example plugins, but it did not support sending the
information to the server data function.



Maybe this will help. This is the code that links datatables to a proxy
function in our backend with the ultimate end point being elastic search.

Can a plugin be written that will achieve the same interaction with the
backend server data function?
If so, I will refactor my code into a plugin.

Thanks,

Chris


/*globals Perimeter,Bodybuilder,moment */
(function () {
    'use strict'; // Enforces browser checking of good java scripting.
Applies to the function that contains it.

    var nsManager = new Perimeter.NamespaceManager(),
        DEVICEREPORTS = nsManager.getNamespace('Perimeter.DeviceReports');


    DEVICEREPORTS.setElasticFnServerData = function (type,
indice_pattern, deviceCustomString2) {

        var recordType = type,
            service = deviceCustomString2;

        this.setFnServerData(function (sSource, aoData, fnCallback, oSettings) {
            var body,
                builder,
                draw = oSettings.iDraw,
                length = oSettings._iDisplayLength,
                start = oSettings._iDisplayStart,
                globalSearch = oSettings.oAjaxData.search.value,
                *globalExclude = oSettings.oAjaxData.exclude.value,*
                columnSearches = aoData[1].value,
                dataSet = [],
                totalHits,
                searches = [],
                key,
                sortDir = [],
                sortCol,
                sortColName = [],
                is_ip_field,
                is_port_field,
                is_date_field,
                is_time_field,
                range,
                from = null,
                to = null,
                values = null,
                ranges = {},
                devices,
                i,
                myFrom,
                longFrom,
                myTo,
                longTo,
                globalTerms
                ;
            devices = Object.keys(DEVICEREPORTS.devices).map(function
(value){ return value.toLowerCase();});
            sortColName.concat('_score');
            sortDir.concat('desc');
            if (aoData[2].value.length > 0) {
                sortCol = aoData[2].value[0].column;
                sortDir.concat(aoData[2].value[0].dir);
                is_ip_field =
oSettings.aoColumns[sortCol].mData.indexOf('Address') > -1;
                is_date_field =
oSettings.aoColumns[sortCol].mData.indexOf('Date') > -1;
                is_time_field =
oSettings.aoColumns[sortCol].mData.indexOf('Time') > -1;

                sortColName.concat((is_ip_field || is_date_field ||
is_time_field) ? oSettings.aoColumns[sortCol].mData :
oSettings.aoColumns[sortCol].mData + '.exact');
            }


            columnSearches.forEach(function (item, index) {

                if (item.search.value.length > 0 && typeof
item.search.value === "string") {
                    is_ip_field = item.data.indexOf('Address') > -1;
                    is_date_field = item.data.indexOf('Date') > -1;
                    is_port_field = item.data.indexOf('Port') > -1;
                    if (is_ip_field || is_port_field) {
                        searches[item.data] = item.search.value;
                    } else if (is_date_field) {
                        values = item.search.value.split(" to ");
                        from = values[0];
                        to = values[1];
                    } else {
                        searches[item.data + ".exact"] = item.search.value;
                    }
                }
            });

            if (from === null) {
                from = moment().subtract(30, 'days').format("YYYY-MM-DD");
            }
            if (to === null) {
                to = moment().format("YYYY-MM-DD");
            }
            ranges.gte = from;
            ranges.lte = to;


            builder = new Bodybuilder().size(length).from(start);
            builder.query("terms", "deviceHostName.exact", devices);
            if ( from.length === 10 && to.length === 10){
                builder.query("range", "deviceCustomDate1", {"lte":
to, "gte": from});

            } else {
                myFrom = new Date(from);
                longFrom = myFrom.getTime()/1000.0;
                myTo = new Date(to);
                longTo = myTo.getTime()/1000.0;
                builder.query("range", "endTime", {"lte": longTo,
"gte": longFrom});

            }


            if (service.constructor === Array) {
                builder.query("terms", "deviceCustomString2", service);
            } else {
                builder.query("match", "deviceCustomString2", service);
            }

            if (globalSearch && globalSearch.length > 0) {

                if (DEVICEREPORTS.validateIpAddress(globalSearch)) {
                    builder.filter("match", "_all", globalSearch);
                } else {
                    globalTerms = globalSearch.toLowerCase().split(' ');
                    if (globalTerms.length > 1){
                        builder.filter("terms", "_all", globalTerms);
                    } else {
                        builder.filter("wildcard", "_all",
globalSearch.toLowerCase() + '*');
                    }
                }
            }

            // if (globalExclude && globalExclude.length > 0) {
            //
            //     if (DEVICEREPORTS.validateIpAddress(globalExclude)) {
            //         builder.filter("match", "_all", globalExclude);
            //     } else {
            //         globalExcludeTerms =
globalExclude.toLowerCase().split(' ');
            //         if (globalExcludeTerms.length > 1){
            //             builder.filter("terms", "_all", globalExcludeTerms);
            //         } else {
            //             builder.filter("wildcard", "_all",
globalExclude.toLowerCase() + '*');
            //         }
            //     }
            // }
            //
            if (Object.size(searches) > 0) {
                for (key in searches) {
                    if (searches.hasOwnProperty(key)) {
                        is_ip_field = key.indexOf('Address') > -1;
                        is_port_field = key.indexOf('Port') > -1;
                        if (is_ip_field) {
                            if
(DEVICEREPORTS.validateIpAddress(searches[key])) {
                                builder.filter("term", key,
searches[key]); // IP column search
                            } else {
                                range =
DEVICEREPORTS.validatePartialIpAddress(searches[key]);
                                if (range !== false) {
                                    builder.filter("range", key,
{"gte": range[0], "lte": range[1]});
                                }
                            }
                        } else if (is_port_field) {
                            builder.filter("term", key,
searches[key]); // Numeric column Search
                        } else {
                            if (typeof searches[key] !== "function") {
                                builder.filter("wildcard", key, "*" +
searches[key] + "*"); // Column Search
                            }
                        }
                    }
                }
            }
            if (sortColName.length > 0) {
                for (i = 0; i < sortColName.length; i += 1) {
                    builder.sort(sortColName[i], sortDir[i]);
                }
            }
            body = builder.build("v2");
            // body.explain = true;
            oSettings.jqXHR =
$.ajax(Routing.generate('ElasticSearchBundle_msearch') + '/' +
recordType, {
                data: JSON.stringify(body),
                type: 'POST',
                contentType: 'application/json',
                success: function (response, textStatus, jqXHR) {
                    if (response.message) {
                        DEVICEREPORTS.showDeviceDataError(response.message);
                        return;
                    }
                    totalHits = response.hits.total <= 10000 ?
response.hits.total : 10000;
                    response.hits.hits.forEach(function (hit) {
                        dataSet.push(hit._source);
                    });
                    fnCallback({
                        "draw": draw,
                        "recordsTotal": totalHits,
                        "recordsFiltered": totalHits,
                        "data": dataSet
                    });
                },
                error: function (jqXHR, textStatus, errorThrown ) {
                    // $('.spinner-mask').modalSpinner('off');
                    if (jqXHR.readyState === 4) {
                        $('#device_report_table_processing').hide();
                    }
                }
            });

        });
    };


}());




  <https://mailtrack.io/> Sent with Mailtrack
<https://mailtrack.io/install?source=signature&lang=en&referral=christopherppinto@gmail.com&idSignature=22>
<#>

On Mon, Apr 10, 2017 at 3:22 PM, Allan Jardine <notifications@github.com>
wrote:

> Thanks for the pull request - much appreciated!
>
> There are quite a lot of white space changes so I wouldn't be able to take
> it as it is. That also makes the diff needlessly large and its hard to see
> where the actual changes are. My guess is that your editor is automatically
> changing tabs to spaces.
>
> I'm also not clear why this needs to be in DataTables core and not added
> as a plug-in. The feature plug-ins
> <https://datatables.net/manual/plug-ins/features> should provide
> everything needed for this. You can register the DOM element and a filter
> plug-in that would perform the exclude action.
>
> Before reviewing the code any further I'd like to just seek some clarity
> on that point.
>
> Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DataTables/DataTablesSrc/pull/101#issuecomment-293052101>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABIZsIWqh44bk8P7BEb2kt9wvWjeZyCYks5ruoGCgaJpZM4M5EH5>
> .
>



-- 
--

Regards,

Chris

ΜΟΛΩΝ ΛΑΒΕ!

DataTables commented:
> If there is a way to get data to the server function that would be great.

You can use the [`preXhr` event](https://datatables.net/reference/event/preXhr) to access and modify the data object that DataTables will send to the server.

The structure of the plug-in would be to create the `input` element and bind a listener to `preXhr` which would get the value from the input and add it to the object sent to the server.

cpinto-cb commented:
"The structure of the plug-in would be to create the input element and bind
a listener to preXhr which would get the value from the input and add it to
the object sent to the server."

This will work for when col search or global filters inputs fire off a
search() or columns.search().

Is there also an event that I can fire to make the server function be
called as someone is typing in my exclude input?

Thanks so much,

Chris








  <https://mailtrack.io/> Sent with Mailtrack
<https://mailtrack.io/install?source=signature&lang=en&referral=christopherppinto@gmail.com&idSignature=22>
<#>

On Tue, Apr 11, 2017 at 3:05 PM, Allan Jardine <notifications@github.com>
wrote:

> If there is a way to get data to the server function that would be great.
>
> You can use the preXhr event
> <https://datatables.net/reference/event/preXhr> to access and modify the
> data object that DataTables will send to the server.
>
> The structure of the plug-in would be to create the input element and
> bind a listener to preXhr which would get the value from the input and
> add it to the object sent to the server.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DataTables/DataTablesSrc/pull/101#issuecomment-293368713>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABIZsPwgwwXL1iwxpnJxozF61LqzWg1Qks5ru87lgaJpZM4M5EH5>
> .
>



-- 
--

Regards,

Chris

ΜΟΛΩΝ ΛΑΒΕ!

DataTables commented:
Hi Chris,

Yes. If you are using server-side processing (which it sounds like you are) simply call the `draw()` API method. That will cause DataTables to make an Ajax request to get the data, which in turn triggers the `preXhr` event and thus your own function to augment the data.

cpinto-cb commented:
Thanks.



On Wed, Apr 12, 2017 at 10:01 AM, Allan Jardine <notifications@github.com>
wrote:

> Hi Chris,
>
> Yes. If you are using server-side processing (which it sounds like you
> are) simply call the draw() API method. That will cause DataTables to
> make an Ajax request to get the data, which in turn triggers the preXhr
> event and thus your own function to augment the data.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DataTables/DataTablesSrc/pull/101#issuecomment-293586301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABIZsAX6rA32riQ5zSDF4nXkF86Bdiyvks5rvNkegaJpZM4M5EH5>
> .
>



-- 
--

Regards,

Chris

ΜΟΛΩΝ ΛΑΒΕ!

cpinto-cb commented:
Allan,

The code came out much simpler.

Thank you for your keen advice.

Unfortunately the code is not portable because of our delayedInput jQuery
extension that we use.


Thanks again,

Chris Pinto


(function ($) {
    "use strict";
    // Exclude plugin

    $.fn.dataTable.ext.feature.push({
        fnInit: function (settings) {
            var api, id, input, table;
            api = new $.fn.dataTable.Api(settings);
            table = api.table();
            id = api.table().node().id;
            input = $(
                '<div id="exclude_filter" class="dataTables_filter">' +
                ' <label><span class="input-group-addon"><i
class="icon icon-search"></i></span>' +
                '  <input type="search" id="exclude_filter"
class="form-control input-sm" placeholder="Enter Exclude Keyword"
aria-controls="device_report_table">' +
                ' </label>' +
                '</div>'
            );
            $(input).find('input')
                .delayedInput({delay: 500})
                .on('chg.delayedInput.perimeter', function (e) {
                    $(this).addClass('focused');
                    var ajaxState = $.fn.DataTable.settings[0].jqXHR.readyState;

                    /*
                     *   If an AJAX is in progress, ajaxState = 1.
Therefore abort the call before creating a new call
                     */
                    if (ajaxState === 1) {
                        $.fn.DataTable.settings[0].jqXHR.abort();
                    }
                    table.draw();
                });
            $('#' + id)
                .on('preXhr.dt', function (e, settings, data) {
                    var value;
                    value =
$(this).closest('#device_report_table_wrapper').find('input#exclude_filter').val();
                    data.exclude = {
                        'regex': false,
                        'value': value
                    };
                });

            return input; // input element
        },
        cFeature: 'E'
    });

}(jQuery));




  <https://mailtrack.io/> Sent with Mailtrack
<https://mailtrack.io/install?source=signature&lang=en&referral=christopherppinto@gmail.com&idSignature=22>
<#>

On Wed, Apr 12, 2017 at 10:14 AM, Chris Pinto <christopherppinto@gmail.com>
wrote:

> Thanks.
>
>
>
> On Wed, Apr 12, 2017 at 10:01 AM, Allan Jardine <notifications@github.com>
> wrote:
>
>> Hi Chris,
>>
>> Yes. If you are using server-side processing (which it sounds like you
>> are) simply call the draw() API method. That will cause DataTables to
>> make an Ajax request to get the data, which in turn triggers the preXhr
>> event and thus your own function to augment the data.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/DataTables/DataTablesSrc/pull/101#issuecomment-293586301>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABIZsAX6rA32riQ5zSDF4nXkF86Bdiyvks5rvNkegaJpZM4M5EH5>
>> .
>>
>
>
>
> --
> --
>
> Regards,
>
> Chris
>
> ΜΟΛΩΝ ΛΑΒΕ!
>
>
>
>


-- 
--

Regards,

Chris

ΜΟΛΩΝ ΛΑΒΕ!

DataTables commented:
Absolutely awesome. Good to hear you got this working!

cpinto-cb commented:
With your guidance, thanks.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Allan Jardine <notifications@github.com>
Sent: Friday, April 14, 2017 8:38:48 AM
To: DataTables/DataTablesSrc
Cc: Christopher Pinto; Author
Subject: Re: [DataTables/DataTablesSrc] Let me explain our application and why we need this feature. (#101)


Absolutely awesome. Good to hear you got this working!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/DataTables/DataTablesSrc/pull/101#issuecomment-294146852>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABIZsLFUjeiu3elrxkmwDd_MD1jQS5Qvks5rv2jYgaJpZM4M5EH5>.
